pkg/virt-controller/watch/vm.go

```diff
@@ -2899,6 +2899,75 @@ func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
 	vm.Status.VolumeRequests = tmpVolRequests
 }
 
+func filterLiveUpdatableVolumes(lastSeenVMSpec *virtv1.VirtualMachineSpec, vm *virtv1.VirtualMachine) {
+	oldVols := make(map[string]virtv1.Volume)
+	for _, v := range lastSeenVMSpec.Template.Spec.Volumes {
+		// The old volumes which can be updated are only hotpluggable
+		// volumes, such as PVCs and DVs.
+		if v.DataVolume != nil || v.PersistentVolumeClaim != nil {
+			oldVols[v.Name] = v
+		}
+	}
+
+	// Evaluate if any volume has changed or has been added
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		oldVol, ok := oldVols[v.Name]
+		switch {
+		// Changes for hotlpugged volumes are valid
+		case storagetypes.IsHotplugVolume(&v):
+			delete(oldVols, v.Name)
+		// The volume has been freshly added
+		case !ok:
+			return
+		// The volume has changed
+		case !equality.Semantic.DeepEqual(oldVol, v):
+			return
+		default:
+			delete(oldVols, v.Name)
+		}
+	}
+	// Evaluate if any volumes were removed and they were hotplugged volumes
+	for _, v := range oldVols {
+		if !storagetypes.IsHotplugVolume(&v) {
+			return
+		}
+	}
+
+	// Evaluate the disks. They need to remain equal except for the ones
+	// associated to hotplugged volumes
+	oldDisks := storagetypes.GetDisksFromVolumes(&lastSeenVMSpec.Template.Spec)
+	// Remove the disks associated to an hotpluggable volume
+	for _, v := range lastSeenVMSpec.Template.Spec.Volumes {
+		if storagetypes.IsHotplugVolume(&v) {
+			delete(oldDisks, v.Name)
+		}
+	}
+	newDisks := storagetypes.GetDisksFromVolumes(&vm.Spec.Template.Spec)
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		oldDisk, okOld := oldDisks[v.Name]
+		newDisk := newDisks[v.Name]
+		switch {
+		case storagetypes.IsHotplugVolume(&v):
+			delete(oldDisks, v.Name)
+		// Old disk has been removed
+		case !okOld:
+			return
+		// Invalid disk change
+		case !equality.Semantic.DeepEqual(oldDisk, newDisk):
+			return
+		default:
+			delete(oldDisks, v.Name)
+		}
+	}
+	if len(oldDisks) > 0 {
+		return
+	}
+
+	// If we reach this point, then all the volumes either remained equal or are live-updatable
+	lastSeenVMSpec.Template.Spec.Volumes = vm.Spec.Template.Spec.Volumes
+	lastSeenVMSpec.Template.Spec.Domain.Devices.Disks = vm.Spec.Template.Spec.Domain.Devices.Disks
```

Comment:
Do you think it would be better if this function simply returned checked that all changes were to "live update" volumes and if so returned true?  Then the caller would do the above assignment

Reply:
Yes, definitely! Also splitting the validation of the volumes and the disks can simplify the logic